### DISCUSSION TITLE
Fields hint and validation messages

### DISCUSSION THREAD

**Discussion by SRFeesha** [wonderflow-bv/wanda#140] [Q&A] [answered]
### Feature description

The Textfield component in Figma comes with a dedicated hint (which is really convenient!) but the `Textfield` component doesn't support it as a prop

### Describe the solution you'd like

Can we add an optional prop to add the hint text on the TextField component? 

### Describe alternatives you've considered

An alternative would be to let the consumer (dev) handle the stuff by themself. 

### Additional context and information

Adding the hint text as a prop will secure the fact that the hint text is following accessibility rules like `aria-describedby`

### Code of Conduct

- [X] I've searched across open and closed issues for similiar requests

**Comment by equinusocio** [comment]
The hint text is not part of the textfield on purpose. Adding it will cause the textfield to **always** have a bottom blank space where to show the hint...even if the hint is not present and never will. That's why we decided to rely on the consumers based on their use cases.

Doing this change now, and adding a blank space under **every** textfield, may cause many UI bugs. We can also chose a different approach, like using tooltip for hint texts.

Another point is, what is an hint text? is a fixed text to provide additional info or a text that popup after validation?

Here some example of using tooltips for validation texts (not present right now)

![Screenshot 2022-04-28 at 09 50 03](https://user-images.githubusercontent.com/10454741/165706471-29f13748-b942-4d1f-b10f-bb0187d10573.png)

This is an example of providing additional field information using a popover:

![Screenshot 2022-04-28 at 09 50 19](https://user-images.githubusercontent.com/10454741/165706579-4a7dc1ca-3408-4e28-98d8-6ce131841b69.png)

**Comment by alessio-libardi** [comment]
If I may add my two cents:

1. Having a Figma component support something and a React component support something else isn't the best way to manage our DS because then we have to create snowflakes where the use case is common since designers have the possibility to create that use case for us dev to bring it to life, either we drop the `hint` from the figma component or we add the prop to the react component;

2. We should support a way to show errors related to an input, for example for required inputs or when some checks don't pass for an input. I think this is related to the hint property because defining the way we manage and display hints and errors underneath the Textfield would improve our consistency throughut the product.

3. There are many DS that are already doing it, we could take upon their foundation to improve our DS, for example Material does it, and I think what they are doing is having a small white space underneath the input to place their hint and errors.

**Reply by equinusocio** [reply]
Providing a `hint` property doesn't solve points 1 and 2, as you can still pass custom elements as hints, consistency is not determined by having this prop, but by constraint and awareness.

Anyway, the example below show how to handle hint texts right now, and it's not considered a snowflake, just normal DS usage:

```jsx
<div>
 <Textfield label="Sample label" />
 <Text as="span" sentimennt="negative" size={14}>Erroring hint text</Text>
</div>
```

So right now you can solve almost every design situation.

> here are many DS that is already doing it, we could take upon their foundation to improve our DS, for example, Material does it, and I think what they are doing is having a small white space underneath the input to place their hint and errors.

Material Design v2 (not the upcoming v3) and most of all DS text fields cause layout shifting, since they don't reserve a fixed space under the field, they just push down everything when the text pops in.

- [Material Design Demo
](https://material.io/components/text-fields#usage
)
- [Carbon Demo](https://carbondesignsystem.com/components/text-input/usage/)

By the way, I think the ways to proceed are two:

- Adding a generic prop to the `Textfield`, which doesn't relate to consistency, doesn't speed up usage but adds a fixed blank space below EVERY `textfield` and `select`.
- Find different solutions.

**Comment by ghost** [comment]
Helper text should only be used when truly necessary. The component should be equipped to support this, it is part of the text field **anatomy**. The helper text ensures that content contains unambiguous labels or instructions that enable users to avoid errors when input is required. Not only that, this will make the component even more accessible by default and we get an improved experience for our users. Tooltips are better suited for context or background information, plus users should not have to look around for a tooltip to access information. An extra spacing should not be the reason why a helper text should not be added.

**Reply by equinusocio** [reply]
> Tooltips are better suited for context
Exactly, they are a really good way to provide field-related and contextual error messages, but not for text hints of course.

Currently, we have many many ways to add support texts to a text field. It is not correct to say "isn't supported"

> An extra spacing should not be the reason why a helper text should not be added.

The extra fixed space is potentially breaking and uncontrollable, but we cannot get rid of it to avoid [layout shifting](https://web.dev/optimize-cls/):

> Layout shifts can be distracting to users. Imagine you've started reading an article when all of a sudden elements shift around the page, throwing you off and requiring you to find your place again. This is very common on the web, including when reading the news, or trying to click those 'Search' or 'Add to Cart' buttons. Such experiences are visually jarring and frustrating. They're often caused when visible elements are forced to move because another element was suddenly added to the page or resized.

**Reply by ghost** [reply]
I understand. If we do use a helper text, spacing should already be determined in the layout, it isn't just going to appear and shift the content down. The helper text is always there and replaced by an error text if invalid. We either use it or not, so there should not be a layout shift.

**Reply by equinusocio** [reply]
Exactly, but the space should have fixed height, so the hint text must no wrap and be really, really short. Check next comment for all the pro and cons

**Reply by alessio-libardi** [reply]
I'm in line with what @Lizfernan is saying. Also, why constraint and awareness are on us dev to tackle if the issue is completely visual, we should contraint and make designers aware of this kind of issue (layout shifting) and how to properly use `hint` and `error` messages when needed.

**Reply by equinusocio** [reply]
Do you have no constraints right now.

> we should contraint and make designers aware of this kind of issue (layout shifting) and how to properly use hint and error messages when needed.

yes that's the point. The issue is not having or not a hint property.

**Comment by equinusocio** [comment]
## Current implementation.

With this implementation, designers are able to put everything below the text field (icons, buttons, HTML elements, other components). No limitations.
These are the usage example comparing now vs. prop:

How to add hint text right now:

```jsx
<div>
 <Textfield label="Sample label" />
 <Text as="span" sentiment="negative" size={14}>Erroring hint text</Text>
</div>
```

By adding an extra prop:

```jsx
<Textfield
  label="Sample label"
  hint={
    <Text as="span" sentiment="negative" size={14}>Erroring hint text</Text>
  }
/>
```

The above example would produce the same result. The hint text must be dynamic since you can provide many different types of data inside (HTML, other components, colored texts). 

#### Pro:

- No empty space if the hint is not present.
- We can use any type of content as hint.
- Hint can be longer and wrap into multi-lines
 
#### Cons:

- Layout shifting occurs whenever a hint is added consequently. So not good for validation messages.

## Fixed space implementation

```jsx
<Textfield
  label="Sample label"
  hint="Hint text"
/>
```

#### Pro:

- Doesn't cause layout shifting

#### Cons:

- Fixed space below every text field and select even if the hint is not present. Same on figma.
- Hint must be really short, it cannot wrap into multi-line since it doesn't take space (absolute positioning).
- Hint can be only a string. No extra element, links, HTML, components etc..

**Reply by alessio-libardi** [reply]
If I may add another cent to the discussion:

As a dev, working to a product with the DS, I don't really care how the issue of displaying hint and error messages are visually tackled, what I care about is:
- what designers prototype is easily reproducible using the DS ✅ 
- what designers prototype doesn't need a snowflake for basic use cases ❌ 

Hint and error messages are a basic use case for input fields. Having them on the DS would solve consistency issues, where one hint message has `dimmed={7}` and `size={14}` and another one has `dimmed={6}` and `weight={'bold'}` or where one error message is red using `<Text sentiment='danger' />` and another one using custom css `color: token(--color-red-80)`.

On a side note there's the issue of making `Textfield` hint and errors accessible. Having them be done on a use case basis would make this definitely harder to achieve, managing them on the DS could highly benefit accessibility matters.

And I know accessibility and consistency can be tackled creting a `CustomTextfield` component on our product but that would really delete the benefits of having a DS.

I would move constraint and awareness on designers to build products that are consistent with the decision taken on the DS to tackle this kind of challenge.

**Reply by equinusocio** [reply]
> Hint and error messages are a basic use case for input fields. Having them on the DS would solve consistency issues, where one hint message has dimmed={7} and size={14} and another one has dimmed={6} and weight={'bold'} or where one error message is red using <Text sentiment='danger' /> and another one using custom css color: token(--color-red-80).

This is true but there are no react properties that solve this issue. As said above: consistency is not determined by having this prop, but by **constraint** and **awareness**, in this specific case awareness. THIS is the only issue I see here since technically speaking there are no inconsistencies in what the designer can do compared to the component. It seems a common case of "less constraints = issues".

> On a side note there's the issue of making Textfield hint and errors accessible. Having them be done on a use case basis would make this definitely harder to achieve, managing them on the DS could highly benefit accessibility matters.

Can you share an example of a11y issue related to the `Textfield`?

> I would move constraint and awareness on designers to build products that are consistent with the decision taken on the DS to tackle this kind of challenge.

The DS involves design and development, so both sides must be covered, even with compromises on one or both sides.

**Comment by equinusocio** [comment, marked as answer]
As discussed with @SRFeesha @Lizfernan and @alessio-libardi we agreed that the `hint` property isn't really required right now unless we redesign how the hint text work (fixed space vs current layout shift), which is unrelated to this discussion.

Proceeding locking the conv.
